### PR TITLE
Feature/recover

### PR DESCRIPTION
**Pull request by themanforfree**
Please briefly answer these questions:
based on #191 
* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)
recover the memory data structure of each store
* what changes does this pull request make?
read persistent storage and recover kv index, permission cache and lease collection
* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)
no

### CODE REVIEW

File: xline/src/storage/auth_store/store.rs [hunk 1]
```diff
@@ -1147,11 +1147,23 @@ where
     pub(crate) fn root_token(&self) -> Result<String, ExecuteError> {
         self.assign(ROOT_USER)
     }
+
+    /// Recover data from persistent storage
+    pub(crate) fn recover(&self) -> Result<(), ExecuteError> {
+        let enabled = self.backend.get_enable()?;
+        if enabled {
+            self.enabled.store(true, AtomicOrdering::Release);
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply xline/src/storage/auth_store/store.rs]: Fixed

**Review by Phoenix500526 — DISMISSED** [review]

File: xline/src/storage/revision.rs [hunk 2]
```diff
@@ -36,13 +36,24 @@ impl Revision {
         self.revision
     }
 
+    // TODO
```
Comment (rogercloud) [inline xline/src/storage/revision.rs:39]: TODO what?

File: xline/src/lib.rs [hunk 3]
```diff
@@ -128,8 +128,26 @@
...
+    allow(
+        clippy::integer_arithmetic,
+        clippy::indexing_slicing,
+        unused_results,
+        clippy::unwrap_used,
+        clippy::str_to_string,
+        clippy::panic,
+        clippy::wildcard_enum_match_arm,
+        clippy::match_wildcard_for_single_variants,
+        clippy::all,
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply xline/src/lib.rs:140]: just copy from curb before, removed unnecessary lints now

File: xline/src/storage/auth_store/backend.rs [hunk 4]
```diff
@@ -116,14 +116,14 @@ where
     }
 
     /// get auth revision
-    pub(crate) fn get_revision(&self) -> Result<RevisionNumber, ExecuteError> {
+    pub(crate) fn get_revision(&self) -> Result<i64, ExecuteError> {
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply xline/src/storage/auth_store/backend.rs:119]: revision's type is `i64`, if use fast path, server will return -1 to client as a invalid revision, and the revision type of etcd is also i64

File: xline/src/storage/auth_store/store.rs [hunk 5]
```diff
@@ -1147,11 +1147,23 @@ where
     pub(crate) fn root_token(&self) -> Result<String, ExecuteError> {
         self.assign(ROOT_USER)
     }
+
+    /// Recover data from persistent storage
+    pub(crate) fn recover(&self) -> Result<(), ExecuteError> {
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply xline/src/storage/auth_store/store.rs:1152]: Fixed

File: xline/src/storage/index.rs [hunk 6]
```diff
@@ -204,4 +214,21 @@ impl IndexOperate for Index {
             new_rev
         }
     }
+
+    fn restore(
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply xline/src/storage/index.rs:218]: Added

File: xline/src/storage/revision.rs (hunk 2 above)
Comment (themanforfree) [reply xline/src/storage/revision.rs:39]: Removed

**Review by Phoenix500526 — APPROVED** [review]

**Review by markcty — APPROVED** [review]

**Pull request merged**